### PR TITLE
Migrate Arch packaging to GH runners

### PR DESCRIPTION
**Pull request by ehildenb**
~Blocked on: https://github.com/runtimeverification/k/pull/3045~

- Parameterize the `test-package` action in the OS and the DISTRO, so that it works for non-debian based distros.
- Allow overriding the dockerfiles to use for `test-package` action.
- Add necessary Dockerfiles in `package/arch` for the `test-package` action to work with it.
- Move the scripts `src/main/scripts/test-in-container*` to `package/test-package` and `package/debian/test-package` scripts.
- Pass in the `test-package` and `build-package` script as parameters to `test-package` github action, so that it can be different for different OSs.
- Remove the Arch packaging from the Jenkins job.

### CODE REVIEW

**Comment by ehildenb** [comment]
It can't find where the packages are moved to, with teh `pkg-name` variable: 
image

Same problem on both Jammy and Arch, so it's something to do with the changes to `test-package` and `with-docker`. Teh `build-package` scripts are supposed to move the packages into place.

**Comment by F-WRunTime** [comment]
> It can't find where the packages are moved to, with teh `pkg-name` variable: image
> 
> Same problem on both Jammy and Arch, so it's something to do with the changes to `test-package` and `with-docker`. Teh `build-package` scripts are supposed to move the packages into place.

Curious. If I recall the package we want is the one coming out of the test-pr job? It could be it's running on two different runners and not being shared as expected via the filesystem? 

Could we look to utilize the upload-artifact and download-artifact which are temporary storage and usable between jobs. 
Upload: https://github.com/marketplace/actions/upload-a-build-artifact
Download: https://github.com/marketplace/actions/download-a-build-artifact

**Comment by ehildenb** [comment]
No they're not shared between workflows, it's all in the same workflow.

File: .github/actions/test-package/action.yml [hunk 1]
```diff
@@ -10,43 +11,64 @@ inputs:
...
+      BUILD_PACKAGE: ${{ inputs.build-package }}
+      PKG_NAME: ${{ inputs.pkg-name }}
     run: |
       set -euxo pipefail
-      docker exec -t k-package-build-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c 'mv package/debian debian'
-      docker exec -t k-package-build-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c "mv debian/compat.${BASE_DISTRO} debian/compat"
-      docker exec -t k-package-build-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c "mv debian/control.${BASE_DISTRO} debian/control"
-      docker exec -t k-package-build-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c "mv debian/rules.${BASE_DISTRO} debian/rules"
-      docker exec -t k-package-build-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c 'dpkg-buildpackage'
+      docker exec -t k-package-build-${BASE_OS}-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c "${BUILD_PACKAGE} ${PKG_NAME}"
```
Comment (ehildenb) [inline .github/actions/test-package/action.yml:59]: Here we call the `BUILD_PACKAGE` script, which we assume moves the created package to `PKG_NAME` location.

File: .github/actions/test-package/action.yml [hunk 2]
```diff
@@ -55,32 +77,33 @@ runs:
...
     shell: bash {0}
     env:
+      BASE_OS: ${{ inputs.os }}
       BASE_DISTRO: ${{ inputs.distro }}
+      TEST_PACKAGE: ${{ inputs.test-package }}
+      PKG_NAME: ${{ inputs.pkg-name }}
     run: |
       set -euxo pipefail
-      docker exec -t k-package-test-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c "k-${BASE_DISTRO}/src/main/scripts/test-in-container-debian"
+      docker exec -t k-package-test-${BASE_OS}-${BASE_DISTRO}-${GITHUB_SHA} /bin/bash -c "${TEST_PACKAGE} ./${PKG_NAME}"
```
Comment (ehildenb) [inline .github/actions/test-package/action.yml]: Then here we call `TEST_PACKAGE`, which we tell where the package has been moved to.

File: .github/workflows/test-pr.yml [hunk 3]
```diff
@@ -30,6 +30,30 @@ jobs:
...
+      - uses: actions/checkout@v3
+      - name: 'Build and Test Package'
+        uses: ./.github/actions/test-package
+        with:
+          dockerfile: package/arch/Dockerfile
+          os: archlinux
+          distro: base
+          build-package: package/arch/build-package
+          test-package: package/arch/test-package
+          pkg-name: kframework_arch_x86_64.pkg.tar.zst
```
Comment (ehildenb) [inline .github/workflows/test-pr.yml]: Here we tell it where the package should be moved to when calling the `test-package` action (where `BUILD_PACKAGE` should move it to, and `TEST_PACKAGE` should assume it's at).

File: .github/workflows/test-pr.yml [hunk 4]
```diff
@@ -30,6 +30,30 @@ jobs:
...
+      - name: 'Upload Package to Release'
+        uses: actions/upload-release-asset@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.JENKINS_GITHUB_PAT }}
+        with:
+          asset_content_type: 'application/gzip'
+          asset_name: 'K Framework Arch Package'
+          asset_path: kframework_arch_x86_64.pkg.tar.zst
```
Comment (ehildenb) [inline .github/workflows/test-pr.yml]: And here we   do the asset upload (so this needs to match the prior step).

Comment (F-WRunTime) [reply .github/workflows/test-pr.yml]: Hmm interesting my first thought is to check the cwd during the execution and maybe that's not what we expected. 
My brain is a bit mush at the moment, I can come back tomorrow morning and give it a run through manually to see if something jumps out at me.

**Comment by ehildenb** [comment]
Evidence this works: https://github.com/runtimeverification/k/actions/runs/3597881818/jobs/6060106730

File: Jenkinsfile [hunk 5]
```diff
@@ -35,68 +35,6 @@ pipeline {
         stash name: 'src', includes: "kframework-${env.VERSION}-src.tar.gz"
```
Comment (F-WRunTime) [inline Jenkinsfile:35]: Was the source tar piece of this build process being primarily used as input to the the following steps? Is the rest of the Jenkinsfile still have value to keep?

**Review by F-WRunTime — APPROVED** [review]

File: Jenkinsfile (hunk 5 above)
Comment (ehildenb) [reply Jenkinsfile:35]: No, this is mostly some legacy thing, I don't think it's actually used anymore. Maybe in building the brew package?

**Pull request merged**